src/web/app/pages-student/student-home-page/student-home-page.component.ts

```diff
@@ -94,67 +96,81 @@ export class StudentHomePageComponent implements OnInit {
+            const isAllSessionsPresent: boolean =
+              sortedFss.filter((fs: FeedbackSession) =>
+                sessionsReturned.has(fs.feedbackSessionName)).length
+                === sortedFss.length;
 
-                if (!isAllSessionsPresent) {
-                  this.statusMessageService.showErrorToast(this.allStudentFeedbackSessionsNotReturned);
-                  this.hasCoursesLoadingFailed = true;
-                  return;
-                }
+            if (!isAllSessionsPresent) {
+              this.statusMessageService.showErrorToast(this.allStudentFeedbackSessionsNotReturned);
+              this.hasCoursesLoadingFailed = true;
+              return;
+            }
 
-                for (const fs of sortedFss) {
-                  const isOpened: boolean = fs.submissionStatus === FeedbackSessionSubmissionStatus.OPEN;
-                  const isWaitingToOpen: boolean =
-                    fs.submissionStatus === FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN;
-                  const isPublished: boolean = fs.publishStatus === FeedbackSessionPublishStatus.PUBLISHED;
+            for (const fs of sortedFss) {
+              const isOpened: boolean = fs.submissionStatus === FeedbackSessionSubmissionStatus.OPEN;
+              const isWaitingToOpen: boolean =
+                fs.submissionStatus === FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN;
+              const isPublished: boolean = fs.publishStatus === FeedbackSessionPublishStatus.PUBLISHED;
 
-                  const isSubmitted: boolean = hasRes.hasResponsesBySession[fs.feedbackSessionName];
-                  studentSessions.push({
-                    isOpened, isWaitingToOpen, isPublished, isSubmitted, session: fs,
-                  });
-                }
-              }, (error: ErrorMessageOutput) => {
-                this.hasCoursesLoadingFailed = true;
-                this.statusMessageService.showErrorToast(error.error.message);
+              const isSubmitted: boolean = hasRes.hasResponsesBySession[fs.feedbackSessionName];
+              studentSessions.push({
+                isOpened, isWaitingToOpen, isPublished, isSubmitted, session: fs,
               });
-
-            this.courses.push({ course, feedbackSessions: studentSessions });
-            this.courses.sort((a: StudentCourse, b: StudentCourse) =>
-              ((a.course.courseId > b.course.courseId) ? 1 : -1));
+            }
           }, (error: ErrorMessageOutput) => {
-            this.hasCoursesLoadingFailed = true;
+            courseRef.hasFeedbackSessionsLoadingFailed = true;
             this.statusMessageService.showErrorToast(error.error.message);
           });
-      }
-    }, (e: ErrorMessageOutput) => {
-      this.hasCoursesLoadingFailed = true;
-      this.statusMessageService.showErrorToast(e.error.message);
-    });
+      }, (error: ErrorMessageOutput) => {
+        courseRef.isFeedbackSessionsLoading = false;
```

Comment:
Is changing this to false necessary? Default value is currently true.

Reply:
Yes it is necessary (though not strictly). Because this is the outer subscription, it will not finalize with `isFeedbackSessionsLoading` set to false (that is set as the inner subscription's finalize method). Therefore, if the outer subscription encounters an error (so inner subscription is not executed), it needs to set the loading to false to indicate an end of the loading process.

Without this, the retry button will still be shown properly as that is the outer element of the loading spinner and probably the page will be the same, but with this set of value which is kind of weird to me:
```
courseRef.isFeedbackSessionsLoading == true;
courseRef.hasFeedbackSessionsLoadingFailed == true;
```